topdown/crypto.go

```diff
@@ -160,12 +164,68 @@ func builtinCryptoSha256(a ast.Value) (ast.Value, error) {
 
 func init() {
 	RegisterFunctionalBuiltin1(ast.CryptoX509ParseCertificates.Name, builtinCryptoX509ParseCertificates)
+	RegisterBuiltinFunc(ast.CryptoX509ParseAndVerifyCertificates.Name, builtinCryptoX509ParseAndVerifyCertificates)
 	RegisterFunctionalBuiltin1(ast.CryptoMd5.Name, builtinCryptoMd5)
 	RegisterFunctionalBuiltin1(ast.CryptoSha1.Name, builtinCryptoSha1)
 	RegisterFunctionalBuiltin1(ast.CryptoSha256.Name, builtinCryptoSha256)
 	RegisterFunctionalBuiltin1(ast.CryptoX509ParseCertificateRequest.Name, builtinCryptoX509ParseCertificateRequest)
 }
 
+func getX509CertsFromBytes(bytes []byte) ([]*x509.Certificate, error) {
+	// if the input is not a PEM string, attempt to decode b64
+	if str := string(bytes); !strings.HasPrefix(str, "-----BEGIN CERTIFICATE-----") {
+		var err error
+		bytes, err = base64.StdEncoding.DecodeString(str)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// attempt to decode input as PEM data
+	p, rest := pem.Decode(bytes)
+	if p != nil && p.Type != blockTypeCertificate {
+		return nil, fmt.Errorf("PEM data contains '%s', expected CERTIFICATE", p.Type)
+	}
+	if p != nil {
+		// if PEM decoded as a valid certificate, use its data as the DER input
+		bytes = p.Bytes
+	}
+
+	// check for more certificates in the chain
+	if p != nil && len(rest) > 0 {
+		var p *pem.Block
+		for {
+			p, rest = pem.Decode(rest)
+			if p == nil {
+				// finish when no more PEM data is read
+				break
+			}
+			// reject any data that isn't exclusively certificates
+			if p.Type != blockTypeCertificate {
+				return nil, fmt.Errorf("PEM data contains '%s', expected CERTIFICATE", p.Type)
+			}
+			bytes = append(bytes, p.Bytes...)
+		}
+	}
+
+	return x509.ParseCertificates(bytes)
+}
+
+func marshalX509Certs(certs []*x509.Certificate) (ast.Value, error) {
+	bs, err := json.Marshal(certs)
+	if err != nil {
+		return nil, err
+	}
+
+	var x interface{}
+
+	if err := util.UnmarshalJSON(bs, &x); err != nil {
+		return nil, err
+	}
+
+	return ast.InterfaceToValue(x)
+}
```

Comment:
❓ Do we still need this, given https://github.com/open-policy-agent/opa/pull/3586?